### ISSUE TITLE
Log errors inside ASGI applications using exc_info, not as strings

### ISSUE THREAD

**Issue by daisylb** [Kludex/uvicorn#225]
There's a few places in uvicorn where it seems to be calling `traceback.format_exc()` on an exception produced by the ASGI application being run, and then log the resulting string using `logger.error()`. This is fine when logging to the console, but when logging to tools like Sentry, the parts of the stack trace that are actually important at the bottom get truncated.

Logging using `logger.error(message, exc_info=True)` instead would pass the actual exception through to Sentry and other similar tools, which means we'd not only get to see the full exception, but we'd also get some of the niceties that the tool provides, such as being able to see locals in each stack frame. The console output should be more or less the same.

The specific case we've been hitting is [this one, on `websockets_impl.py`](https://github.com/encode/uvicorn/blob/7977b95eae7a01fab2e390af120eb3dec47fd03e/uvicorn/protocols/websockets/websockets_impl.py#L143), but it looks like there's a couple of other usages of `.format_exc()` where the same pattern is being used too.

**Comment by lovelydinosaur** [comment]
Yes good call. A pull request for this would be appreciated.

What'd also be good as part of that would be to make sure we're not using bare exceptions in those cases, but instead `except BaseException as exc` and pass the exc info to the logger explicitly, rather than just passing a boolean `True`.

This prevents Python implicitly creating a cyclical reference.

**Comment by almarklein** [comment]
This should do the trick: `logger.error(..., exc_info=(type(exc ), exc , exc .__traceback__))`

**Comment by jessebraham** [comment]
I was just working on this issue along with #191, and was looking for a bit of clarification on @almarklein's comment before I make a pull request.

To my understanding, Python 3.5+ allows passing the `exc_info` keyword argument either:
- a Tuple in the format you have provided (ie. what is returned by `sys.exc_info()`  
- the exception instance (ie. `exc`).

Is there a benefit I'm unaware of to providing the Tuple you demonstrated, or should I just provide the exception instance to keep things clean?

Thanks :)

**Comment by almarklein** [comment]
@jessebraham Good catch! Apparently, you can also simply pass the exception object, which leads to much cleaner code. I just dived in Python's `logging` source and found:
```py
            if isinstance(exc_info, BaseException):
                exc_info = (type(exc_info), exc_info, exc_info.__traceback__)
```

**Comment by lovelydinosaur** [comment]
Resolved in 0.3.19
